Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -2805,7 +2826,8 @@ protected class NullClassWriter : IClassWriter {
 
       type = type.NormalizeExpandKeepConstraints();
       Contract.Assert(type is NonProxyType);  // this should never happen, since all types should have been successfully resolved
-      return TypeInitializationValue(type, wr, tok, false, constructTypeParameterDefaultsFromTypeDescriptors);
+      var usePlaceboValue = type.AsDatatype?.GetGroundingCtor().IsGhost == true;
```

Comment:
The precondition of this `DefaultValue` method previously stated `Contract.Requires(type.HasCompilableValue);`, so I was going to answer "yes". But then I experimentally found that this condition does not hold at all call sites. After a further investigation, my conclusion is that the precondition was incorrectly stated. It _is_ fine and expected that this method can be called with a type that has no compilable value. This happens when `type` (or any component of `type`) is a type parameter that is not auto-init. In those cases, `DefaultValue` emits code that produces a default value either from a parameter in the target-code context (for example, this may be a parameter `G _default_G` that gives a `G` value to be used as default) or from a type descriptor in the target-code context (for example, the target-code context may contain a parameter `Dafny.TypeDescriptor<G> _td_G` and the default value will then be produced from `_td_G.Default()`).

My conclusion is that the precondition of `DefaultValue` was stated incorrectly. So, I removed it.